src/transactions/TransactionFrame.cpp

```diff
@@ -142,27 +143,51 @@ TransactionFrame::checkSignature(AccountFrame& account, int32_t neededWeight)
     // calculate the weight of the signatures
     int totalWeight = 0;
 
-    for (size_t i = 0; i < getEnvelope().signatures.size(); i++)
+    // compare all available SIGNER_KEY_TYPE_HASH_TX with current transaction hash
+    // current transaction hash is not stored in getEnvelope().signatures - it is
+    // computed with getContentsHash() method
+    for (auto it = keyWeights.begin(); it != keyWeights.end(); it++)
     {
-        auto const& sig = getEnvelope().signatures[i];
-
-        for (auto it = keyWeights.begin(); it != keyWeights.end(); it++)
+        auto &signerKey = it->key;
+        if (signerKey.type() == SIGNER_KEY_TYPE_HASH_TX)
         {
-            if (PubKeyUtils::hasHint((*it).pubKey, sig.hint) &&
-                PubKeyUtils::verifySig((*it).pubKey, sig.signature,
-                                       contentsHash))
+            if (signerKey.hashTx() == contentsHash)
             {
-                mUsedSignatures[i] = true;
+                mUsedOneTimeSignerKeys.insert(signerKey);
```

Comment:
This mUsedOneTimeSignerKeys also should be cleared at the start of TransactionFrame::apply    since checkSignature is also called when a tx first comes in when deciding to propagate or drop it.